src/tls.c

```diff
@@ -1579,134 +1579,190 @@ static int TLSX_SetALPN(TLSX** extensions, const void* data, word16 size,
-                     ssl->heap);
-    if (r != WOLFSSL_SUCCESS) {
-        WOLFSSL_MSG("TLSX_SetALPN failed");
-        return BUFFER_ERROR;
+cleanup:
+    *pextension = extension;
+    *psel = sel;
+    *psel_len = sel_len;
+    return r;
+}
+
+int ALPN_Select(WOLFSSL *ssl)
+{
+    TLSX *extension;
+    const byte *sel = NULL;
+    byte sel_len = 0;
+    int r = 0;
+
+    WOLFSSL_ENTER("ALPN_Select");
+    if (ssl->alpn_peer_requested == NULL) {
+        goto cleanup;
     }
 
-    /* reply to ALPN extension sent from client */
-    if (isRequest) {
+#if defined(OPENSSL_ALL) || defined(WOLFSSL_NGINX) || defined(WOLFSSL_HAPROXY)
+    if (ssl->alpnSelect != NULL && ssl->options.side == WOLFSSL_SERVER_END) {
+        if (ssl->alpnSelect(ssl, &sel, &sel_len, ssl->alpn_peer_requested,
+                            ssl->alpn_peer_requested_length,
+                            ssl->alpnSelectArg) == 0) {
+            WOLFSSL_MSG_EX("ALPN protocol match");
+        }
+        else {
+            sel = NULL;
+            sel_len = 0;
+        }
+    }
+#endif
+
+    if (sel == NULL) {
+        r = ALPN_find_match(ssl, &extension, &sel, &sel_len,
+                            ssl->alpn_peer_requested,
+                            ssl->alpn_peer_requested_length);
+        if (r != 0)
+            goto cleanup;
+    }
+
+    if (sel != NULL) {
+        /* set the matching negotiated protocol */
+        r = TLSX_SetALPN(&ssl->extensions, sel, sel_len, ssl->heap);
+        if (r != WOLFSSL_SUCCESS) {
+            WOLFSSL_MSG("TLSX_SetALPN failed");
+            r = BUFFER_ERROR;
+            goto cleanup;
+        }
+        /* reply to ALPN extension sent from peer */
 #ifndef NO_WOLFSSL_SERVER
         TLSX_SetResponse(ssl, TLSX_APPLICATION_LAYER_PROTOCOL);
 #endif
     }
+    r = 0;
 
-    return 0;
+cleanup:
+    WOLFSSL_LEAVE("ALPN_Select", r);
+    return r;
+}
+
+/** Parses a buffer of ALPN extensions and set the first one matching
+ * client and server requirements */
+static int TLSX_ALPN_ParseAndSet(WOLFSSL *ssl, const byte *input, word16 length,
+                                 byte isRequest)
+{
+    word16  size = 0, offset = 0, wlen;
+    int     r = BUFFER_ERROR;
+    TLSX    *extension;
+    const byte *s;
+
+    if (OPAQUE16_LEN > length)
+        goto cleanup;
+
+    ato16(input, &size);
+    offset += OPAQUE16_LEN;
+
+    /* validating alpn list length */
+    if (size == 0 || length != OPAQUE16_LEN + size)
+        goto cleanup;
+    /* validating length of entries before accepting */
+    for (s = input + offset, wlen = 0; (s - input) < size; s += wlen) {
+        wlen = *s++;
+        if (wlen == 0 || (s + wlen - input) > length)
+            goto cleanup;
+    }
+
+    if (isRequest) {
+        /* keep the list sent by peer, if this is from a request. We
+         * use it later in ALPN_Select() for evaluation. */
+        if (ssl->alpn_peer_requested != NULL)
+            XFREE(ssl->alpn_peer_requested, ssl->heap, DYNAMIC_TYPE_ALPN);
```

Comment:
XFREE checks for NULL, so you can remove the check on line 1723.

Reply:
Added the reset of `ssl->alpn_peer_requested_length` to `0` here also.